syntax = "proto3";

package common.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/any.proto";

option go_package = "github.com/vitalstream/protos/common/v1;commonv1";
option java_package = "com.vitalstream.common.v1";
option java_multiple_files = true;

// Health check request
message HealthCheckRequest {
  string service = 1;
}

// Health check response
message HealthCheckResponse {
  enum ServingStatus {
    SERVING_STATUS_UNSPECIFIED = 0;
    SERVING_STATUS_SERVING = 1;
    SERVING_STATUS_NOT_SERVING = 2;
    SERVING_STATUS_SERVICE_UNKNOWN = 3;
  }
  ServingStatus status = 1;
  string message = 2;
  google.protobuf.Timestamp timestamp = 3;
}

// Request metadata
message RequestMetadata {
  string request_id = 1;
  string user_id = 2;
  string session_id = 3;
  string client_version = 4;
  map<string, string> custom_headers = 5;
  google.protobuf.Timestamp timestamp = 6;
}

// Response metadata
message ResponseMetadata {
  string request_id = 1;
  string server_version = 2;
  google.protobuf.Duration processing_time = 3;
  map<string, string> custom_headers = 4;
  google.protobuf.Timestamp timestamp = 5;
}

// Empty message
message Empty {
}

// UUID message
message UUID {
  string value = 1;
}

// Coordinate message
message Coordinate {
  double latitude = 1;
  double longitude = 2;
  double altitude = 3;
}

// Address message
message Address {
  string street = 1;
  string city = 2;
  string state = 3;
  string country = 4;
  string postal_code = 5;
  Coordinate coordinates = 6;
}

// Phone number message
message PhoneNumber {
  string number = 1;
  string country_code = 2;
  string type = 3; // mobile, home, work, etc.
}

// Email message
message Email {
  string address = 1;
  string type = 2; // personal, work, etc.
  bool primary = 3;
}

// Person name message
message PersonName {
  string prefix = 1;
  string first_name = 2;
  string middle_name = 3;
  string last_name = 4;
  string suffix = 5;
}

// Timestamp range
message TimestampRange {
  google.protobuf.Timestamp start = 1;
  google.protobuf.Timestamp end = 2;
}

// Duration range
message DurationRange {
  google.protobuf.Duration min = 1;
  google.protobuf.Duration max = 2;
}

// Numeric range
message NumericRange {
  double min = 1;
  double max = 2;
}

// Integer range
message IntegerRange {
  int32 min = 1;
  int32 max = 2;
}

// File info
message FileInfo {
  string name = 1;
  string mime_type = 2;
  int64 size = 3;
  string checksum = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
}

// Version info
message VersionInfo {
  string version = 1;
  string build = 2;
  string commit = 3;
  google.protobuf.Timestamp build_time = 4;
}

// Environment info
message EnvironmentInfo {
  string name = 1;
  string region = 2;
  string zone = 3;
  map<string, string> variables = 4;
}

// Service info
message ServiceInfo {
  string name = 1;
  string version = 2;
  string description = 3;
  repeated string endpoints = 4;
  map<string, string> metadata = 5;
}

// Metric value
message MetricValue {
  string name = 1;
  double value = 2;
  google.protobuf.Timestamp timestamp = 3;
  map<string, string> labels = 4;
  string unit = 5;
}

// Log entry
message LogEntry {
  enum LogLevel {
    LOG_LEVEL_UNSPECIFIED = 0;
    LOG_LEVEL_DEBUG = 1;
    LOG_LEVEL_INFO = 2;
    LOG_LEVEL_WARNING = 3;
    LOG_LEVEL_ERROR = 4;
    LOG_LEVEL_FATAL = 5;
  }
  
  LogLevel level = 1;
  string message = 2;
  string service = 3;
  string component = 4;
  string trace_id = 5;
  string span_id = 6;
  google.protobuf.Timestamp timestamp = 7;
  map<string, string> metadata = 8;
}

// Tag
message Tag {
  string key = 1;
  string value = 2;
  string description = 3;
}

// Label
message Label {
  string key = 1;
  string value = 2;
}

// Annotation
message Annotation {
  string key = 1;
  string value = 2;
}

// Selector
message Selector {
  repeated Label match_labels = 1;
  repeated Label match_expressions = 2;
}

// Condition
message Condition {
  string field = 1;
  string operator = 2;
  string value = 3;
}

// Filter
message Filter {
  repeated Condition conditions = 1;
  string operator = 2; // AND, OR, NOT
}

// Sort
message Sort {
  string field = 1;
  enum Direction {
    DIRECTION_UNSPECIFIED = 0;
    DIRECTION_ASC = 1;
    DIRECTION_DESC = 2;
  }
  Direction direction = 2;
}

// Search query
message SearchQuery {
  string text = 1;
  repeated string fields = 2;
  Filter filter = 3;
  repeated Sort sort = 4;
  int32 limit = 5;
  int32 offset = 6;
}

// Validation rule
message ValidationRule {
  string field = 1;
  string rule = 2;
  string message = 3;
  bool required = 4;
  string pattern = 5;
  repeated string allowed_values = 6;
  NumericRange numeric_range = 7;
  IntegerRange integer_range = 8;
  DurationRange duration_range = 9;
  TimestampRange timestamp_range = 10;
}

// Configuration
message Configuration {
  string key = 1;
  string value = 2;
  string type = 3;
  string description = 4;
  bool encrypted = 5;
  google.protobuf.Timestamp updated_at = 6;
  string updated_by = 7;
}

// Feature flag
message FeatureFlag {
  string name = 1;
  bool enabled = 2;
  string description = 3;
  repeated string conditions = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  string updated_by = 7;
}
